File: sim3/sim.cpp
#include <iostream>
using namespace std;

class Node {
    public:
        Node *next = nullptr;
        Node *previous = nullptr;
        char character;
        Node(char c) {
            this->character = c;
        }
};

class LinkedList {
    private:
        int currentSize = 0;
    public:
        Node *head = nullptr;
        Node *tail = nullptr;
        void append(char c) {
            if (this->currentSize == 0) {
                head = new Node(c);
                tail = head;
            }
            else {
                Node* current = new Node(c);
                tail->next = current;
                current->previous = tail;
                this->currentSize++;
            }
        }
        char pop() {
            char c = tail->character;
            Node *old = tail;
            tail = tail->previous;
            tail->next = nullptr;
            this->currentSize--;
            delete old;
            return c;
        }
        void join(LinkedList* endList) {
            tail->next = endList->head;
            endList->head->previous = tail;
            tail = endList->tail;
            this->currentSize += endList->size();
            //delete endList;
        }
        void print() {
            int length = this->size();
            char* output = new char[length+1];
            Node* current = this->head;
            for (int i=0; i<length; i++) {
                output[i] = current->character;
                current = current->next;
            }
            output[length] = '\n';
            //cout << output <<endl;
            printf("%s", output);
            delete[] output;
        }
        int size() {
            return this->currentSize;
        }
        //in life, may want to delete list while passing items to another one. 
        /*
        ~LinkedList() {
            //this->deleteItems();
        }*/
        void deleteItems() {
            if (this->currentSize > 0) {
                Node *current = this->head;
                Node *next = this->head;
                while (current != nullptr) {//finished with tail. 
                    next = current->next;
                    delete current;
                    current = next;
                }
            }
        }
};

LinkedList* parse(const string& str) {
    LinkedList* front = new LinkedList();
    LinkedList* mainList = new LinkedList();
    LinkedList* current = mainList;
    for (char c : str) {
        if (c == '<') {
            if (current->size() > 0) {
                current->pop();
            }
        }
        else if (c == ']') {
            if (front->size() > 0) {
                front->join(mainList);
                mainList = front;
                front = new LinkedList();
            }
            current = mainList;
        }
        else if (c == '[') {
            if (front->size() > 0) {
                front->join(mainList);
                mainList = front;
                front = new LinkedList();
            }
            current = front;
        }
        else {
            current->append(c);
        }
    }
    if (front->size() > 0) {
        front->join(mainList);
        mainList = front;
    }
    else {
        delete front;//check for memory leaks.
    }
    return mainList;
}

//will need to modify this to accept input. 
int main() {
    int lines;
    char extra;
    char *output = nullptr;
    string line;
    LinkedList* list;
    //getline(cin, lines);
    cin >> lines;
    cin >> extra;
    for (int i=0; i<lines; i++) {
        getline(cin, line);//read in input line.
        cout << line << endl;
        list = parse(line);//read line.  
        list->print();
    }
    delete list;
    return 0;
}

File: sim3/sim.py
from collections import deque
import timeit

class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.previous = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0
    def append(self, c):
        if self.size > 0:
            self.tail.next = Node(c)
            tail = self.tail
            self.tail = self.tail.next
            self.tail.previous = tail
        else:
            self.tail = Node(c)
            self.head = self.tail
        self.size += 1
    def pop(self):
        if self.size > 0:
            self.tail = self.tail.previous
            self.tail.next = None
            self.size -= 1
    def join(self, list2):
        self.append(list2.head)
        self.size += (list2.size-1)
    def print(self):
        current = self.head
        output = ""
        while (current != None):
            output += current.value
            current = current.next
        print(output)     


def parse(string):
    front = LinkedList()
    main = LinkedList()
    current = main
    for char in string:
        if char == '<':
            if current.size > 0:
                current.pop()
        elif char == ']':
            if front.size > 0:
                front.join(main)
                main = front
                front = LinkedList()
            current = main
        elif char == '[':
            if front.size > 0:
                front.join(main)
                main = front
                front = LinkedList()
            current = front
        else:
            current.append(char)
    if front.size > 0:
        front.join(main)
        main = front
    return main

#this takes too long.
#would do c extension but must be one file. 
def dequeToString(mydeque):
    str = ""
    for char in mydeque:
        str += char
    return str

def main():
    test_cases = int(input())
    #outputs = deque([])
    for test in range(test_cases):
        string = input()
        mylist = parse(string)
        mylist.print()
        #for char in list:
            #print(char, end='')

main()
